src/pymoca/tree.py

```diff
@@ -259,6 +273,694 @@ def handle_walk(self, listener: TreeListener, tree: Union[ast.Node, dict, list])
+    5. Qualified Import names, see 4 (but not from Extends Clauses) (spec 13.2.1)
+    6. Public Unqualified Imports (error if multiple are found) (spec 13.2.1)
+    7. Repeat 1-6 for each lexically enclosing instance scope, stopping at `encapsulated`
+    unless predefined type, function, operator. If name matches a variable (a.k.a. component,
+    symbol) in an enclosing class, it must be a `constant`.
+    """
+
+    current_scope = scope
+    while True:
+        if (
+            (
+                found := _find_local(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                )
+            )
+            or (
+                found := _find_inherited(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or search_imports
+            and (
+                found := _find_imported(
+                    name,
+                    current_scope,
+                    check_builtin_classes=check_builtin_classes,
+                    current_extends=current_extends,
+                )
+            )
+            or not search_parent
+            or not current_scope.parent
+            or current_scope.encapsulated
+        ):
+            break
+        current_scope = current_scope.parent
+    # If name matches a variable (a.k.a. component a.k.a. symbol) in an enclosing class,
+    # it must be a `constant`.
+    if (
+        isinstance(found, ast.Symbol)
+        and current_scope != scope
+        and "constant" not in found.prefixes
+    ):
+        raise NameLookupError("Non-constant Symbol found in enclosing class")
+
+    # If not found and we stopped at an encapsulated class,
+    # then search predefined functions and operators in global scope
+    # TODO: Add predefined functions and operators to global scope before this
+    if found is None and current_scope.encapsulated:
+        found = _find_local(name, scope.root)
+        if not isinstance(found, ast.Class) or found.type not in ("function", "operator"):
+            found = None
+
+    return found
+
+
+def _find_rest_of_name(
+    first: Union[ast.Class, ast.Symbol], rest_of_name: str
+) -> Optional[Union[ast.Class, ast.Symbol]]:
+    """Lookup the `B.C` part of Composite Name Lookup (`A.B.C`) (spec 5.3.2)
+
+    1. `A` is looked up using Simple Name Lookup and passed as `first` argument
+    2. If `A` is a Component:
+        1. `B.C` is looked up from named component elements of `A`
+        2. if not found and if `A.B.C` is used as a function call and `A` is a scalar or can be
+        evaluated as a scalar from an array and `B` and `C` are classes,
+        it is a non-operator function call.
+    3. If `A` is a Class:
+        1. `A` is temporarily flattened without modifiers
```

Comment:
I don't know. What I have in there now passes the `ModelicaCompliance.Scoping.NameLookup` tests without any flattening. Maybe there are other ModelicaCompliance cases that specifically get to this issue and would fail. Ignorance is bliss until you crash and burn :laughing:.

I don't see where the spec says *why* "temporary flattening" is required. Maybe when I get into the flattening it will become more apparent.

Reply:
@jackvreeken I tried your example. 

Using current flattening and `Class.find_class` it fails with
`pymoca.ast.ClassNotFoundError: Could not find class 'A.B'`.
Using current flattening with `tree.find_name`, the first try fails with
`pymoca.tree.NameLookupError: A is not a package so B must be encapsulated`.
Making B `encapsulated`, it passes giving `M.bla=1`. Trying `find_class` with the `encapsulated` fix still fails.

So the answer to your question is lookup was the problem.

Man, you have an impressive ability to come up with good examples to break things! 👍 🥇